### ISSUE TITLE
length_fn compile error 

### ISSUE THREAD

**Issue by charloco** [libpnet/libpnet#112]
I have the following test struct

#[packet]
pub struct MsgTest{

    source: u16be,

    destination: u16be,

    #[length_fn="msg_options_length"]

    options:Vec<u8>,

    #[payload]
    payload: Vec<u8>
}
this works well.But when I add "    t:u8," after options:Vec<u8> just as following codes:
pub struct MsgTest{

    source: u16be,

    destination: u16be,

    #[length_fn="msg_options_length"]

    options:  Vec <u8>,

    t:u8, //this cause compile error

    #[payload]
    payload: Vec<u8>
}

the error message is 

<quote expansion>:141:47: 141:58 error: use of moved value: `self.packet` [E0382]
<quote expansion>:141         self_.packet[4 + msg_options_length(&self.to_immutable()) + 0] = (val) as u8;

**Comment by mrmonday** [comment]
It's hard to tell you what the issue is you're encountering without seeing the code that's causing it - by the looks of things it's certainly not an issue with libpnet.

There's more information about the error you're getting, and how to fix it, here: http://doc.rust-lang.org/nightly/error-index.html#E0382

If you post a more complete version of your code I'd be happy to help you fix it.

**Comment by charloco** [comment]
in fact,i am trying to write a mqtt server with libpnet.So i do some test first.

I create a libpnet/src/packet/mqtt.rs

and give the test struct 

```
#[packet]
pub struct Mqtt {
    source: u16be,
    destination: u16be,
    //length: u16be,
   // checksum: u16be,
    #[length_fn="mqtt_options_length"]
    options:Vec<u8>,
    t:u8,
    #[payload]
    payload: Vec<u8>
}
```
I guess that between "length_fn" and "payload" any field should not insert,

or can say that length_fn cann't be placed any place.

the full error msg

```
<quote expansion>:141:47: 141:51 error: use of moved value: `*self` [E0382]
<quote expansion>:141         self_.packet[4 + mqtt_options_length(&self.to_immutable()) + 0] = (val) as u8;
                                                                    ^~~~
/home/yuchang/work/git/rust/libpnet/src/packet/mqtt.rs:31:1: 31:10 note: in expansion of #[_packet_generator]
/home/yuchang/work/git/rust/libpnet/src/packet/mqtt.rs:31:1: 31:10 note: expansion site
<quote expansion>:141:47: 141:51 help: run `rustc --explain E0382` to see a detailed explanation
<quote expansion>:140:13: 140:18 note: `self` moved here because it has type `&mut packet::mqtt::MutableMqttPacket<'a>`, which is moved by default
<quote expansion>:140         let self_ = self;
                                  ^~~~~
/home/yuchang/work/git/rust/libpnet/src/packet/mqtt.rs:31:1: 31:10 note: in expansion of #[_packet_generator]
/home/yuchang/work/git/rust/libpnet/src/packet/mqtt.rs:31:1: 31:10 note: expansion site
<quote expansion>:140:13: 140:18 help: if you would like to borrow the value instead, use a `ref` binding as shown:
<quote expansion>:            let ref self_ = self;
/home/yuchang/work/git/rust/libpnet/src/packet/mqtt.rs:31:1: 31:10 note: in expansion of #[_packet_generator]
/home/yuchang/work/git/rust/libpnet/src/packet/mqtt.rs:31:1: 31:10 note: expansion site
error: aborting due to previous error
Could not compile `pnet`.
```

and the "self_.packet[4 + mqtt_options_length(&self.to_immutable()) + 0]" comes from decorator.rs

```
 979 fn generate_sop_strings(offset: &str, operations: &[SetOperation]) -> String {
 980     let mut op_strings = String::new();
 981     for (idx, sop) in operations.iter().enumerate() {
 982         let pkt_replace = format!("self_.packet[{} + {}]", offset, idx); ///here 
 983         let val_replace = "val";
 984         let sop = sop.to_string().replace("{packet}", &pkt_replace[..])
 985                                  .replace("{val}", val_replace);
 986         op_strings = op_strings + &sop[..] + ";\n";
 987     }
 988 
 989     op_strings
 990 }
```

**Comment by mrmonday** [comment]
Pull #113 should hopefully fix this.

**Comment by mrmonday** [comment]
Fixed in master, and in 0.2.0 of pnet and pnet_macros crates on crates.io :)

**Comment by charloco** [comment]
Thanks :smiley:
